### PR TITLE
fix: eliminate hover-driven style recalc storm in thread list

### PR DESCRIPTION
**Pull request by XyneSpaces**
## What
Fixes the main-thread CPU pin (~91%) when hovering messages in a long thread. Applies fixes #1 and #2 from the performance profile; toolbar-remount (#3) and thread virtualization are intentionally left for a follow-up.

## Root cause
Each `pointerover` forced a synchronous style recalc that re-evaluated **every** animatable property on every `transition-all` element, starting thousands of `scrollbar-color` transitions (99% of the storm) — a pure `transition-all` artifact. The hover toolbar compounded it by reading `getBoundingClientRect()` twice per event, flushing layout synchronously.

## Changes
**#1 — Scope animations to what actually changes**
- `ChatBubble` bubble root: `transition-all` → `transition-colors` (only `data-[hovered]:bg-muted/50` animates). This is the amplifier — one instance per rendered message.
- `MessageAttachment` card: `transition-all` → `transition-shadow` (preserves the intended `hover:shadow-md`).
- `MessageAttachment` hover overlay: `transition-all` → `transition-colors`.

**#2 — Remove the forced synchronous reflow per hover** (`MessageHoverToolbar`)
- Cache the container's viewport rect; recompute lazily only after scroll/resize instead of reading `getBoundingClientRect()` twice per `pointerover`.
- Read row geometry **before** stamping `data-hovered`, so the layout read no longer flushes the style invalidation it just created.

## Intentionally not touched
- `ExpandableMessage` / `RenderMessageWithHTML` keep `transition-all` — they animate `max-height`/spacing for expand-collapse, not color.
- `MessageHoverToolbar` `key={hoverKey}` remount (#3) and `ThreadList` virtualization are separate follow-ups (this will likely be merged with the thread-virtualization work).

## Validation
- `pnpm --filter dashboard typecheck` passes clean.
- Manual re-profile of thread hover still pending (please re-profile after merge/preview).

## Notes
No XYNE ticket key was provided for the branch/title — happy to rename if there's a ticket to attach.

### CODE REVIEW

**Review by santhoshJus — APPROVED** [review]

File: apps/dashboard/src/components/Chat/HoverActionsToolbar/MessageHoverToolbar.tsx [hunk 1]
```diff
@@ -90,6 +90,16 @@ export const MessageHoverToolbar: React.FC<MessageHoverToolbarProps> = ({ contai
     const container = containerRef.current;
     if (!container) return;
 
+    // The container's viewport rect only shifts on scroll/resize/layout, not
+    // per hovered row. Reading it on every pointerover forced a second
+    // synchronous layout flush per event; cache it and recompute lazily after
+    // a scroll/resize instead.
+    let containerTop = 0;
```
Comment (XyneSpaces) [inline apps/dashboard/src/components/Chat/HoverActionsToolbar/MessageHoverToolbar.tsx:97]: **[🟠 High] R7 — new cache + invalidation logic ships with no test.** The dirty-flag lifecycle introduced here (scroll/resize invalidation, lazy recompute, read-before-stamp ordering, pinned freeze) is new behavior validated only by `typecheck`, and the PR description itself notes the manual re-profile is still pending. Before merge, add a component test (hover positions the toolbar at the row; scroll invalidates and repositions; pinned popover freezes it) or land the re-profile evidence — otherwise a regression in this path (e.g. the stale-cache case flagged at the recompute site) is invisible to CI. (Review-guidelines rule 204: unit tests for new functionality and edge cases.)

File: apps/dashboard/src/components/Chat/HoverActionsToolbar/MessageHoverToolbar.tsx [hunk 2]
```diff
@@ -117,16 +127,22 @@ export const MessageHoverToolbar: React.FC<MessageHoverToolbarProps> = ({ contai
       const hoverKey = row.getAttribute('data-hover-key');
       if (!messageId || !hoverKey) return;
 
+      // Read layout geometry BEFORE stamping data-hovered so this
+      // getBoundingClientRect read does not flush the style invalidation we are
+      // about to create. The container rect is cached across rows.
+      if (containerTopDirty) {
```
Comment (XyneSpaces) [inline apps/dashboard/src/components/Chat/HoverActionsToolbar/MessageHoverToolbar.tsx:133]: **[🟡 Medium] R2 — stale `containerTop` on ancestor layout shifts.** The cache is invalidated only by scroll (capture) and window `resize`. If an ancestor of the container changes layout without either firing — e.g. a banner/call strip inserted or collapsing above the thread pane, or a sibling pane drag-resize (assumption: such ancestors exist in the thread pane; they're outside this diff) — `containerTop` stays stale and the toolbar renders offset by the shift until the next scroll/resize self-heals it. Suggestion: bound the staleness instead of caching across events — recompute at most once per animation frame (max one forced layout per frame), or additionally mark dirty from a `ResizeObserver` on the container's parent (note: RO on the container itself won't catch position-only moves).

File: apps/dashboard/src/components/ui/MessageBubble/MessageBubble.tsx [hunk 3]
```diff
@@ -1158,7 +1158,7 @@ export const MessageBubble: React.FC<MessageBubbleProps> = ({
                           }
                         }
                   }
-                  className={`${isMobile ? 'text-[12px]' : 'text-xs'} text-muted-foreground cursor-pointer hover:underline transition-all duration-150 visual-regression-hide ${searchItemView ? 'ml-auto shrink-0' : ''}`}
+                  className={`${isMobile ? 'text-[12px]' : 'text-xs'} text-muted-foreground cursor-pointer hover:underline transition-colors duration-150 visual-regression-hide ${searchItemView ? 'ml-auto shrink-0' : ''}`}
```
Comment (XyneSpaces) [inline apps/dashboard/src/components/ui/MessageBubble/MessageBubble.tsx:1161]: **[🟡 Medium] R9 — PR description doesn't list this file.** This commit adds three `transition-all` → `transition-colors` narrowings here (timestamp link, reaction chips, emoji-picker trigger), but the description's "Changes" section still lists only `ChatBubble` and `MessageAttachment`. The changes themselves are consistent with the fix — all three only animate color properties on hover, so `transition-colors` is correct — just update the description so it matches the diff. (Review-guidelines rule 233 rates this 🟠; downgraded here since it's doc-only and trivially fixed.)

**Review by XyneSpaces — COMMENTED** [review]
Review pass on head `de35a99` (4 files, 9 hunks — all checked). The core fix is sound: `setHighlightedRow` stamps `data-hovered` imperatively, so the read-before-stamp reorder genuinely avoids the forced synchronous flush, the listener cleanup correctly matches capture flags, and this commit's removal of the duplicate `mouseover`/`mouseleave` pair eliminates a real double-fire per mouse move (pointer events cover mouse/touch/pen — verified safe). Two fix-before-merge findings and two low/verification notes inline; full findings table and reviewer recommendation in the top-level summary comment. Posting as COMMENT only — no approve/request-changes gate.

File: apps/dashboard/src/components/Chat/ChatBubble/ChatBubble.tsx [hunk 4]
```diff
@@ -1162,7 +1162,7 @@ export const ChatBubble: React.FC<ChatBubbleProps> = ({
       data-hover-key={hoverToolbarKey}
       className={cn(
         isMobile && 'no-select-mobile',
-        'group/bubble relative transition-all duration-200 ease-in-out',
+        'group/bubble relative transition-colors duration-200 ease-in-out',
```
Comment (XyneSpaces) [inline apps/dashboard/src/components/Chat/ChatBubble/ChatBubble.tsx:1165]: **[🟢 Low] R8 — confirm the bubble root only animates colors.** `transition-colors` is correct if the only animated property on this root is `data-[hovered]:bg-muted/50`, as the description states. The remainder of this `cn()` call is outside the diff context, so I couldn't verify there's no ring/shadow/transform hover class on the root that would now snap instead of fading — a quick confirm would close this out. (Verified for the `MessageAttachment` card: its `focus:ring-*` utilities are box-shadow-based, so `transition-shadow` covers them.)

**Comment by XyneSpaces** [comment]
## Xyne Spaces Reviewer — PR #1547

**Verdict:** Approve-worthy after fixes — no blocking issues, but address the 1 High + 2 Medium findings before merge.
**Reviewed:** 4 files, 9 hunks @ head `de35a99` · **Findings:** 0 blocking, 1 high, 2 medium, 1 low

Worth noting as done well: the read-before-stamp reorder is correct (`setHighlightedRow` writes `data-hovered` imperatively, so the old order really did force a synchronous flush per hover), the listener cleanup matches capture flags, and this commit's removal of the duplicate `mouseover`/`mouseleave` pair eliminates a genuine double-fire per mouse move. Migration safety: no migration paths touched — nothing to gate.

### Findings
| Severity | Rule | Location | Issue |
|----------|------|----------|-------|
| 🟠 high | R7 (rule 204) | `MessageHoverToolbar.tsx:97` | New cache/invalidation logic ships with no test; validation is typecheck-only and the manual re-profile is pending |
| 🟡 medium | R2 | `MessageHoverToolbar.tsx:133` | `containerTop` cache invalidated only on scroll/resize — ancestor layout shifts leave the toolbar mispositioned until the next scroll |
| 🟡 medium | R9 (rule 233) | `ui/MessageBubble.tsx:1161` | Description's Changes section doesn't list this file's 3 new transition narrowings |
| 🟢 low | R8 | `ChatBubble.tsx:1165` | Confirm the bubble root has no non-color hover animation that now snaps (remainder of the `cn()` call is outside diff context — assumption) |

### Recommended reviewer
**@satvikbatra** — please add him via the GitHub UI (Reviewers → `satvikbatra`): this agent session has no reviewer-request API available, so the request could not be posted programmatically. This @mention notifies him in the meantime.

| Candidate | Owns (files in this PR) | CODEOWNER? | Pending reviews | Tie-breaker |
|-----------|-------------------------|-----------|-----------------|-------------|
| @satvikbatra | ChatBubble.tsx + ChatListV4 (the toolbar's container) via 6 reviews on #1393; closest human to MessageHoverToolbar.tsx (no human PR history on it) | no | n/a* | 1 active Spaces ticket; active today |
| @harshpreet931 | MessageAttachment.tsx (authored #414/#415/#439/#527, #1262 hover card) | no | n/a* | 5 active tickets incl. 1 CRITICAL — heaviest load |
| @Dnyaneshwari-04 | ChatBubble.tsx (authored #1398) | no | n/a* | 4 STARTED tickets |
| @Mohan0544 | chat-area reviewer (#1398, #920) | no | n/a* | 0 open tickets, but coordinating release unblock today |
| @devesh-juspay | approved #1393 | no | n/a* | 3 STARTED tickets; security-focused work lately |

\* The `review-requested:` GitHub search qualifier returned the unfiltered open-PR set for every candidate, so per-person pending-review counts aren't computable from search; the PRs inspected directly (#1547, #1539) have no requested reviewers. Workload was therefore ranked on Spaces ticket load + recent activity. No CODEOWNERS entry matches the changed `apps/dashboard` paths (CODEOWNERS covers ACL paths only), so ranking is file-ownership + workload.

**Rationale:** @satvikbatra reviewed the most recent change to this exact subsystem (#1393 — ChatBubble plus the ChatListV4 container that hosts the shared hover toolbar) across six review submissions on Sep 2–3, has the lightest active load among the people familiar with these files, and was active in Spaces today. @santhoshJus, who profiled the original hover-recalc issue, is already engaged as the requester.

File: apps/dashboard/src/components/Chat/HoverActionsToolbar/MessageHoverToolbar.tsx [hunk 5]
```diff
@@ -90,6 +90,16 @@ export const MessageHoverToolbar: React.FC<MessageHoverToolbarProps> = ({ contai
     const container = containerRef.current;
     if (!container) return;
 
+    // The container's viewport rect only shifts on scroll/resize/layout, not
+    // per hovered row. Reading it on every pointerover forced a second
+    // synchronous layout flush per event; cache it and recompute lazily after
+    // a scroll/resize instead.
+    let containerTop = 0;
+    let containerTopDirty = true;
```
Comment (XyneSpaces) [inline apps/dashboard/src/components/Chat/HoverActionsToolbar/MessageHoverToolbar.tsx:98]: **[minor] R2** — `containerTop` is invalidated only by a scroll inside the container or a window `resize`; any other container shift (a banner/pillar appearing above the list, an outer-ancestor scroll, a side-panel toggle changing the container's offset) leaves the cache stale, so the next `pointerover` positions the toolbar off by that shift, and it stays wrong until the next scroll/resize. Suggested fix: also mark the cache dirty from a `ResizeObserver` on the container and/or a window-level `scroll` listener (capture) so outer scrolls invalidate it.

File: apps/dashboard/src/components/Chat/HoverActionsToolbar/MessageHoverToolbar.tsx [hunk 6]
```diff
@@ -117,16 +127,22 @@ export const MessageHoverToolbar: React.FC<MessageHoverToolbarProps> = ({ contai
...
       if (!messageId || !hoverKey) return;
 
+      // Read layout geometry BEFORE stamping data-hovered so this
+      // getBoundingClientRect read does not flush the style invalidation we are
+      // about to create. The container rect is cached across rows.
+      if (containerTopDirty) {
+        containerTop = container.getBoundingClientRect().top;
+        containerTopDirty = false;
+      }
+      const top = Math.round(row.getBoundingClientRect().top - containerTop);
```
Comment (XyneSpaces) [inline apps/dashboard/src/components/Chat/HoverActionsToolbar/MessageHoverToolbar.tsx:137]: **[major] R7** — this is a behavior-changing fix (container-rect caching + geometry-read-before-`data-hovered`-stamp + listener consolidation) with no test added, and the PR description's validation section says the post-fix manual re-profile is still pending. Add a small component test for the cache lifecycle (scroll/resize mark it dirty; same-row repeat short-circuits via `activeRowRef`), or attach the re-profile results before merge.

File: apps/dashboard/src/components/Chat/HoverActionsToolbar/MessageHoverToolbar.tsx [hunk 7]
```diff
@@ -138,23 +154,25 @@ export const MessageHoverToolbar: React.FC<MessageHoverToolbarProps> = ({ contai
...
     const handleScroll = (): void => {
+      markContainerDirty();
       if (pinnedOpenRef.current) return;
       hide();
     };
 
+    // Pointer events already cover mouse, touch and pen, so one
+    // pointerover/pointerleave pair is enough. The old mouseover/mouseleave
+    // duplicates fired handlePointerOver twice per mouse move.
     container.addEventListener('pointerover', handlePointerOver);
```
Comment (XyneSpaces) [inline apps/dashboard/src/components/Chat/HoverActionsToolbar/MessageHoverToolbar.tsx:165]: **[minor] R5** — dropping the `mouseover`/`mouseleave` pair removes the contract that open PR #292 (arrow-key message navigation) relies on: its PR body says selection moves dispatch a synthetic `mouseover` on the target row to drive this toolbar. If #292 is still planned, either keep a `mouseover` bridge until it lands or update it to dispatch a synthetic `pointerover`. *Assumption: #292 is still active — confirming with its author would settle this.*

**Review by XyneSpaces — COMMENTED** [review]
Automated Xyne Spaces review — 4 findings (0 blocking, 1 major, 3 minor) across 4 files / 9 hunks at `de35a998`. Details anchored to the changed lines below; a full summary comment follows. No migration files touched, so the migration-safety gate does not apply.

File: apps/dashboard/src/components/ui/MessageBubble/MessageBubble.tsx (hunk 3 above)
Comment (XyneSpaces) [inline apps/dashboard/src/components/ui/MessageBubble/MessageBubble.tsx:1161]: **[minor] R9** — the PR description's Changes section enumerates the other `transition-all` → scoped-transition changes but omits this file's three (timestamp, reaction chip, emoji-add button); update the description so it matches the diff (guidelines rule 233).

**Comment by XyneSpaces** [comment]
## Xyne Spaces Reviewer — PR #1547

**Verdict:** Changes requested — nothing blocking (0 blocking, 1 major, 3 minor)
**Reviewed:** 4 files, 9 hunks (+31/−13) @ `de35a998` · **Findings:** 1 major, 3 minor

### Findings
| Severity | Rule | Location | Issue |
|----------|------|----------|-------|
| major | R7 | `apps/dashboard/src/components/Chat/HoverActionsToolbar/MessageHoverToolbar.tsx:137` | Behavior-changing positioning fix ships with no test; the PR's own validation (manual re-profile) is still pending |
| minor | R2 | `apps/dashboard/src/components/Chat/HoverActionsToolbar/MessageHoverToolbar.tsx:98` | `containerTop` cache is invalidated only by container scroll + window resize; other layout shifts leave the toolbar offset until the next scroll |
| minor | R5 | `apps/dashboard/src/components/Chat/HoverActionsToolbar/MessageHoverToolbar.tsx:165` | Removing the `mouseover`/`mouseleave` pair drops the synthetic-`mouseover` contract that open PR #292 (arrow-key nav) depends on |
| minor | R9 | `apps/dashboard/src/components/ui/MessageBubble/MessageBubble.tsx:1161` | This file's 3 transition changes are missing from the PR description's Changes list (rule 233) |

No migration files touched — migration-safety review not applicable.

### Recommended reviewer
**@Dnyaneshwari-04** — owns the closest recent work in this PR's components (drove #914 which changed `MessageAttachment.tsx`; assignee of XYNE-61728 scoped to `ChatBubble`/`ThreadList` message-list components; merged dashboard PR #1543 today) and has 0 pending review requests. Note: this bot's GitHub toolset exposes no review-request API (and no `gh` CLI), so the formal reviewer request could not be filed — this @mention is the tag; please also add her via the reviewers sidebar for the formal entry.

| Candidate | Owns / recent work in this PR's components | CODEOWNER match? | Pending reviews | Tie-breaker |
|-----------|-------------------------------------------|------------------|-----------------|-------------|
| @Dnyaneshwari-04 | `MessageAttachment.tsx` (#914 driver), `ChatBubble`/`ThreadList` scope (XYNE-61728), merged #1543 (dashboard) today | no (CODEOWNERS covers ACL paths only) | 0 | most recent activity + deepest component overlap |
| @harshpreet931 | approved #914 (`MessageAttachment.tsx`), broader chat-context work | no | 0 | — |
| @santhoshJus | drove the perf analysis behind this PR; knows all 4 files | no | 0 | excluded — already submitted an approving review on this PR |

Signals not computable with this bot's tools: per-file commit history (`/commits?path=`) is not exposed, so file ownership was proxied via PR search over the changed components + Spaces context; CODEOWNERS covers only ACL paths, so it yields no eligibility set for these files. Repo-wide pending-review load is 0 — none of the ~30 open PRs has a requested reviewer.

**Comment by XyneSpaces** [comment]
@Dnyaneshwari-04 — requesting your review on this dashboard chat hover-recalc fix.

You’re the strongest mappable reviewer signal: you recently authored `ChatBubble.tsx` changes in PR #1040, and this PR touches `ChatBubble.tsx`, `MessageHoverToolbar.tsx`, and `MessageAttachment.tsx` in the same hover/action path.

Note: the GitHub MCP toolset available here does not expose the requested-reviewers endpoint, so I can’t set the Reviewers field directly. If this mention doesn’t register as a formal review request, please add yourself via the GitHub PR sidebar.
